Update nRF54H20 soc and board definitions to be IronSide compatible

jonathannilsen opened the pull request:
This PR updates the nRF54H20 soc and board definitions to be compatible with IronSide SE
and the new nRF54H20 software architecture, replacing the old definitions.

Included are changes from the following PRs, as well as some additional adaptations:
* https://github.com/zephyrproject-rtos/zephyr/pull/91667
* https://github.com/zephyrproject-rtos/zephyr/pull/91413
* https://github.com/zephyrproject-rtos/zephyr/pull/91005

Note that currently there are some manual steps required to construct the PERIPHCONF
necessary to run samples/tests that need a non-standard global domain peripheral config,
for example setting CTRLSEL for fast peripheral pins.

jonathannilsen committed:
west: runners: nrf: Program UICR/PERIPHCONF artifacts if present

Program the new UICR and PERIPHCONF artifacts if they are generated.
These are required for the application to operate properly if they
are in use.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

57300 committed:
modules: hal_nordic: Remove nrf-regtool support for nRF54H

nrf-regtool will not be used as part of IronSide SE compatible builds.
It will remain in use for the nRF92 series, until that too undergoes a
switch from SDFW to IronSide SE.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
boards: nrf54h20dk: Merge iron variants into the base variants

This replaces the legacy SDFW compatible board configuration with the
IronSide SE compatible one, thus removing support for running samples
and tests on nRF54H20 devices with the old firmware.

All applications are expected to work on `nrf54h20dk/nrf54h20/cpuapp`
out of the box. For other board targets, all applications are expected
to boot, but may require additional peripheral configuration in UICR.
Build system support for the new UICR format is to be added separately.

Co-authored-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>
Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>
Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
boards: nrf54h20dk: Limit app core FLASH_LOAD_SIZE

CONFIG_USE_DT_CODE_PARTITION had to be disabled to add MCUboot support.
As a result, CONFIG_FLASH_LOAD_SIZE was left at zero, which means that
the linker would claim all available MRAM for the app core.

For now, we can't allow that, because the default nRF54H20 DK memory map
divides MRAM between multiple cores in order to support various samples.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

jonathannilsen committed:
boards: nordic: update nrf54h20dk memory map

Update memory map to be compatible with latest IronSide.

180kB MRAM is reserved.

Co-authored-by: Håkon Amundsen <haakon.amundsen@nordicsemi.no>
Signed-off-by: Håkon Amundsen <haakon.amundsen@nordicsemi.no>
Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

jonathannilsen committed:
dts: nordic: update UICR definition on nrf54h20

With IronSide SE there is only one defined UICR which is at
the location of the APPLICATION UICR. Update the devicetree
definition accordingly, and use the "nordic,nrf-uicr" compatible
on the node since the domain distinction added by the v2 compatible
is no longer relevant.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

jonathannilsen committed:
soc: nordic: add IronSide SE compatible UICR support

Add support for generating UICR and associated artifacts in a
format compatible with IronSide SE, to be used for Nordic SoCs
in the Haltium family.

The main feature added with this is the ability to configure certain
global domain peripherals that are managed by the secure domain
through setting UICR.PERIPHCONF. This register points at a blob of
(register address, register value) pairs which are loaded
into the peripherals by IronSide SE ahead of the application boot.

The added helper macros in uicr.h can be used to add register
configurations to the PERIPHCONF. Entries added through these macros
are then extracted by a script, post-processed and placed in a blob
located at specific part of MRAM.

A default PERIPHCONF configuration has been added for the nrf54h20
soc to support the standard BLE use case (matching the configuration
in the soc devicetree).

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

jonathannilsen committed:
west: runners: nrf: don't recover twice on nrf54h

Update the recover mechanism for nrf54h to only call recover
once. Using nrfutil device recover with both --core Network and
--core Application is redundant with IronSide SE as both of these
map to the same operation which does a full erase of the device MRAM.

Additionally, recovering twice in a row specifically in a nrfutil
batch file (which is used by this runner implementation) triggers some
odd behavior with the current latest version of
nrfutil device + IronSide SE, which can cause the device to enter a
reset loop and appear unresponsive and preventing 'west flash --recover'
from working properly.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

jonathannilsen committed:
samples: drivers: mbox: switch nrf54h20 cpuapp and cpurad remotes

Switch which board is the remote in the test case for
mbox communication between nrf54h20dk/nrf54h20/cpuapp and
nrf54h20dk/nrf54h20/cpurad, making cpurad the remote instead.
This is done to prepare the sample for executing with IronSide SE,
where using cpurad as the main board doesn't make as much sense,
since cpuapp has to start cpurad.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

karstenkoenig commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-memory_map.dtsi:
```diff
@@ -16,14 +16,6 @@
 			#size-cells = <1>;
 			ranges = <0x0 0x2f010000 0x41000>;
```
We can extend cpuapp_ram0x to the start of 0x2f000000 now. That includes deleting the offset for cpuapp_data defined still below

jonathannilsen committed:
boards: nordic: nrf54h20dk: refactor RAM memory map

Refactor the default RAM memory map on nrf54h20dk:

Removes use of "nordic,owned-memory" which is no longer needed on
nrf54h20. Reserved memory nodes that were under "nordic,owned-memory"
have been moved directly under reserved-memory.

The memory shared between cpuapp-cpusec and cpurad-cpusec in RAM0x
is no longer used with IronSide, since IPC buffers toward the secure
domain are at new fixed locations. The cpuapp_data region
has been expanded to fill the available space in RAM0x when removing
these shared memory regions.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

nordicjm commented on boards/nordic/nrf54h20dk/Kconfig.defconfig:
```diff
@@ -19,6 +19,13 @@ if !USE_DT_CODE_PARTITION
 config FLASH_LOAD_OFFSET
 	default $(dt_nodelabel_reg_addr_hex,cpuapp_boot_partition)
 
+# This is meant to span 'cpuapp_boot_partition' and 'cpuapp_slot0_partition'
+# in the default memory map.
+config FLASH_LOAD_SIZE
+	default $(add_hex, $(dt_nodelabel_reg_addr_hex,cpuapp_slot0_partition), \
+	  $(dt_nodelabel_reg_size_hex,cpuapp_slot0_partition), \
+	  -$(dt_nodelabel_reg_addr_hex,cpuapp_boot_partition))
```
use a subpartition in dts?

hakonfam committed:
soc: nrf54h: don't boot radio core if VTOR is not programmed

Booting the radio core when it is not programmed will typically
cause a reset loop. This can happen when programming multiple
images to a device, and the app core image is programmed before
the radio core.

With this change we avoid the reset loop in that case.

Signed-off-by: Håkon Amundsen <haakon.amundsen@nordicsemi.no>

57300 committed:
tests: boards: nrf: nrfs: Build main image for app core only

Update this multi-core test to always run the `main` and `remote` images
on cpuapp and cpurad respectively.

This is to prepare the test for running with IronSide SE, in which case
keeping cpurad as the main board target wouldn't make as much sense,
because cpurad would have to be started by cpuapp.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

nordicjm commented on soc/nordic/common/uicr/Kconfig:
```diff
@@ -19,6 +19,49 @@ config NRF_HALTIUM_UICR_PERIPHCONF
 	  to point at the blob. The initialization values are then loaded ahead of
 	  ahead of the application boot.
 
+config NRF_HALTIUM_UICR_RECOVERY
+	bool "Configure UICR RECOVERY field"
+	help
+	  Configure the UICR RECOVERY field to enable recovery firmware boot.
+	  This should be enabled by the main application to configure recovery
+	  firmware parameters. The recovery firmware itself does not need to
+	  enable this option.
```
this should be at sysbuild level rather than adding it to every image whereby it's useless for most

carlescufi commented:
DNM before next week so that @SebastianBoe and @57300 can take a look.

57300 commented:
> DNM before next week so that @SebastianBoe and @57300 can take a look.

I reviewed it today @carlescufi, we're just waiting for people to refresh their +1s.
